quantumblacklabs/kedro-viz
# Display warnings on large graphs

## Description

Related ticket: https://jira.quantumblack.com/browse/KED-2280

This ticket is to implement the 'large pipeline warning' user flow, where a warning message will  prompt the user to reselect tags and nodes to allow a smaller pipeline to be displayed.  The user can also choose to proceed with displaying the graph. 

![image](https://user-images.githubusercontent.com/27775658/106582931-73318380-653c-11eb-8243-54d52d3952fc.png)

Update: There is a new update to the design and the code has been updated accordingly as below:

![image](https://user-images.githubusercontent.com/27775658/106787735-fb507f80-6647-11eb-9a3b-040e1558d5c6.png)

The warning will look like the following under the light theme
![image](https://user-images.githubusercontent.com/27775658/107243347-d29fff80-6a24-11eb-9651-041d59d56f47.png)


## Development notes

There are 2 main areas of development for this PR: the set up of the modal, along with the styling; and the set up of a series of state, actions, reducers and selectors that allow the app to determine if the pipeline is a 'large' pipeline before the graph gets rendered. 

These are the main things I have included in this PR:
1. Set up of the modal and its related styling - I have used react-modal for the modal setup, while I have adopted very simple and straight forward styling rules. ( Please do let me know if there are better ways to incorporate inherit styles for the modal)
2. Introducing 4 new fields in the non-pipeline state that governs the display of the modal: ' isLarge', 'DisplayLargeGraph', 'NodesNo' and 'EdgesNo'. Both 'isLarge' is needed to listen to the condition of whether a pipeline is large as a first criteria to trigger the warning, while 'DisplayLargeGraph' is used to listen to the user's preference for whether to display the large pipeline. Both 'NodesNo' and 'EdgesNo' are for quick tally of the no. of Nodes and no. of edges to allow both figures to be displayed without the need for the entire set of nodes and edges to be passed on down to the modal component, thereby allowing the warning to be triggered almost immediately on load. 
3. refactor of actions in `graph.js` to control the display of the warning message without loading the loader and the graph
4. adding in selectors to expose the newly added fields
5. adding in reducers to allow changing the values of the newly added fields
6. Tests for the modal and the newly added actions
7. A field within the config to determine the max combination of nodes and edges to display the warning message

## QA notes

This is best to be tested with a large pipeline, which could be imitated by using the random data genereator. The expected behavior are as follows:
1. On first load, a large pipeline would trigger the warning message. 
2. On selecting different tags, if the combination still results in a large pipeline, the warning message will persist with updated no. of tags and edges in the warning message.
3. On selecting a tag that results in a smaller pipeline, the graph will immediately load.
4. On selecting 'render it anyway', the large pipeline will always render within the lifetime of the session until the app is reloaded. 

## Checklist


## Legal notice


- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
